File: java_solutions_1/236.lowest-common-ancestor-of-a-binary-tree.java
/*
 * @lc app=leetcode id=236 lang=java
 *
 * [236] Lowest Common Ancestor of a Binary Tree
 */

// @lc code=start

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return root == null || root == p || root == q ? root : java.util.stream.Stream.of(0).map(_v -> new TreeNode[]{lowestCommonAncestor(root.left, p, q), lowestCommonAncestor(root.right, p, q)}).map(a -> new TreeNode[]{a[0] != null && a[1] != null ? root : a[0] != null ? a[0] : a[1]}).findFirst().get()[0];
    }
}
// @lc code=end

File: kotlin_solutions/236.lowest-common-ancestor-of-a-binary-tree.kt
/*
 * @lc app=leetcode id=236 lang=kotlin
 *
 * [236] Lowest Common Ancestor of a Binary Tree
 */

class Solution {fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? = root?.let { if (it == p || it == q) it else lowestCommonAncestor(it.left, p, q).let { left -> lowestCommonAncestor(it.right, p, q).let { right -> if (left != null && right != null) root else left ?: right } } }}

